File: demo/for-lua/boolmatrix.lua
local sudoku = require("libluasudoku")

return {
	New = function(rows, cols, maxNodes)
		local matrix = {
			handler = sudoku.CreateBoolMatrix(rows, cols, maxNodes)
		}
		setmetatable(matrix, {
			__index = {
				SetMatrixRowData = function(matrix, row)
					-- here can be optimized
					sudoku.SetMatrixRowData(matrix.handler, row)
				end,
				DancingLinks = function(matrix, justOne, callback)
					return sudoku.DancingLinks(matrix.handler, justOne, callback)
				end
			},
			__gc = function(matrix)
				sudoku.DestroyBoolMatrix(matrix.handler)
			end
		})
		return matrix
	end
}

File: demo/for-php/boolmatrix.php
<?php

class BoolMatrix
{
	private $handler = NULL;
	public static $callback;
	function __construct($rows, $cols, $maxNodes)
	{
		$this->handler = PHPCreateBoolMatrix($rows, $cols, $maxNodes);
	}
	function __destruct()
	{
		printf("Destroy Bool Matrix\n");
		PHPDestroyBoolMatrix($this->handler);
	}
	function SetMatrixRowData($row)
	{
		PHPSetMatrixRowData($this->handler, $row);
	}
	function DancingLinks($justOne, Closure $callback)
	{
		BoolMatrix::$callback = $callback;
		return PHPDancingLinks($this->handler, $justOne);
	}
}

function MatrixAnswerCallback($answer)
{
	if (BoolMatrix::$callback != NULL) {
		call_user_func(BoolMatrix::$callback, $answer);
	}
}
